### PR TITLE
[full-ci]first api test for ocm

### PR DESCRIPTION
**Pull request by ScharfViktor**
related #1329

how to start two ocis instances:

- create providers.json file:

<details>
<summary>  providers.json </summary>

```
[
        {
                "name": "localhost:9200",
                "full_name": "Local",
                "organization": "Owncloud",
                "domain": "localhost",
                "homepage": "https://owncloud.com",
                "services": [
                        {
                                "endpoint": {
                                        "type": {
                                                "name": "OCM",
                                                "description": "CERNBox Open Cloud Mesh API"
                                        },
                                        "name": "CERNBox - OCM API",
                                        "path": "https://localhost:9200/ocm/",
                                        "is_monitored": true
                                },
                                "api_version": "0.0.1",
                                "host": "localhost:9200"
                        }
                ]
        },
        {
                "name": "anotherlocalhost:10200",
                "full_name": "Local 10200",
                "organization": "Owncloud",
                "domain": "anotherlocalhost",
                "homepage": "https://owncloud.com",
                "services": [
                        {
                                "endpoint": {
                                        "type": {
                                                "name": "OCM",
                                                "description": "CERNBox Open Cloud Mesh API"
                                        },
                                        "name": "CERNBox - OCM API",
                                        "path": "https://anotherlocalhost:10200/ocm/",
                                        "is_monitored": true
                                },
                                "api_version": "0.0.1",
                                "host": "anotherlocalhost:10200"
                        }
                ]
        }
]
```

</details>

- put `127.0.0.1       anotherlocalhost` to `/etc/hosts`
- put `ocm` to https://github.com/owncloud/ocis/blob/master/services/web/pkg/config/defaults/defaultconfig.go#L101
- build ocis `make -C ocis clean generate build`
- start first instance https://localhost:9200: `PROXY_ENABLE_BASIC_AUTH=true OCIS_LOG_PRETTY=true IDM_CREATE_DEMO_USERS=true ADMIN_PASSWORD=admin FRONTEND_OCS_INCLUDE_OCM_SHAREES=true FRONTEND_OCS_LIST_OCM_SHARES=true FRONTEND_ENABLE_FEDERATED_SHARING_INCOMING=true FRONTEND_ENABLE_FEDERATED_SHARING_OUTGOING=true OCIS_ADD_RUN_SERVICES=ocm GRAPH_INCLUDE_OCM_SHAREES=true OCM_OCM_INVITE_MANAGER_INSECURE=true OCM_OCM_SHARE_PROVIDER_INSECURE=true OCM_OCM_STORAGE_PROVIDER_INSECURE=true OCM_OCM_PROVIDER_AUTHORIZER_PROVIDERS_FILE=providers.json ocis/bin/ocis server
`
- start second ocis instance https://anotherlocalhost:9200
using launch.json vscode: 

<details>
<summary> launch.json </summary>

```
{
          "name": "10200",
          "type": "go",
          "request": "launch",
          "mode": "debug",
          "program": "${workspaceFolder}/ocis/cmd/ocis/main.go",
          "args": ["server"],
          "env": {
              "OCM_OCM_PROVIDER_AUTHORIZER_PROVIDERS_FILE": "/Users/scharfviktor/Work/ocis/providers.json",
              "OCM_OCM_INVITE_MANAGER_INSECURE": "true",
              "OCM_OCM_SHARE_PROVIDER_INSECURE": "true",
              "OCM_OCM_STORAGE_PROVIDER_INSECURE": "true",
              "FRONTEND_OCS_INCLUDE_OCM_SHAREES": "true",
              "OCIS_BASE_DATA_PATH": "/Users/scharfviktor/.ocis-10200",
              "OCIS_CONFIG_DIR": "/Users/scharfviktor/.ocis-10200/config",
              "OCIS_EVENTS_ENDPOINT": "127.0.0.1:10233",
              "OCIS_LDAP_URI": "ldaps://localhost:10235",
              "OCIS_RUNTIME_PORT": "10250",
              "OCIS_URL": "https://anotherlocalhost:10200",
              // "OCIS_OIDC_ISSUER": "anotherlocalhost",
              "FRONTEND_OCS_LIST_OCM_SHARES": "true",
              "FRONTEND_ENABLE_FEDERATED_SHARING_INCOMING": "true",
              "FRONTEND_ENABLE_FEDERATED_SHARING_OUTGOING":"true",
              "OCIS_ADD_RUN_SERVICES": "ocm",
              "GRAPH_INCLUDE_OCM_SHAREES": "true",
              "APP_PROVIDER_DEBUG_ADDR": "127.0.0.1:10165",
              "APP_PROVIDER_GRPC_ADDR": "127.0.0.1:10164",
              "APP_REGISTRY_DEBUG_ADDR": "127.0.0.1:10243",
              "APP_REGISTRY_GRPC_ADDR": "127.0.0.1:10242",
              "AUTH_BASIC_DEBUG_ADDR": "127.0.0.1:10147",
              "AUTH_BASIC_GRPC_ADDR": "127.0.0.1:10146",
              "AUTH_MACHINE_DEBUG_ADDR": "127.0.0.1:10167",
              "AUTH_MACHINE_GRPC_ADDR": "127.0.0.1:10166",
              "AUTH_SERVICE_DEBUG_ADDR": "127.0.0.1:10169",
              "AUTH_SERVICE_GRPC_ADDR": "127.0.0.1:10199",
              "CLIENTLOG_DEBUG_ADDR": "127.0.0.1:10260",
              "EVENTHISTORY_DEBUG_ADDR": "127.0.0.1:10270",
              "FRONTEND_DEBUG_ADDR": "127.0.0.1:10141",
              "FRONTEND_HTTP_ADDR": "127.0.0.1:10140",
              "GATEWAY_DEBUG_ADDR": "127.0.0.1:10143",
              "GATEWAY_GRPC_ADDR": "127.0.0.1:10142",
              "GRAPH_DEBUG_ADDR": "127.0.0.1:10124",
              "GRAPH_HTTP_ADDR": "127.0.0.1:10120",
              "GROUPS_DEBUG_ADDR": "127.0.0.1:10161",
              "GROUPS_GRPC_ADDR": "127.0.0.1:10160",
              "IDM_DEBUG_ADDR": "127.0.0.1:10239",
              "IDM_LDAPS_ADDR": "127.0.0.1:10235",
              "IDP_DEBUG_ADDR": "127.0.0.1:10134",
              "IDP_HTTP_ADDR": "127.0.0.1:10130",
              "NATS_DEBUG_ADDR": "127.0.0.1:10234",
              "NATS_NATS_PORT": "10233",
              "OCDAV_DEBUG_ADDR": "127.0.0.1:10163",
              "OCM_DEBUG_ADDR": "127.0.0.1:10281",
              "OCM_HTTP_ADDR": "127.0.0.1:10280",
              "OCM_GRPC_ADDR": "127.0.0.1:10282",
              "OCS_DEBUG_ADDR": "127.0.0.1:10114",
              "OCS_HTTP_ADDR": "127.0.0.1:10110",
              "POSTPROCESSING_DEBUG_ADDR": "127.0.0.1:10255",
              "POSTPROCESSING_SERVICE_DEBUG_ADDR": "127.0.0.1:10255",
              "PROXY_DEBUG_ADDR": "127.0.0.1:10205",
              "PROXY_HTTP_ADDR": "127.0.0.1:10200",
              "SEARCH_DEBUG_ADDR": "127.0.0.1:10224",
              "SEARCH_GRPC_ADDR": "127.0.0.1:10220",
              "SETTINGS_DEBUG_ADDR": "127.0.0.1:10189",
              "SETTINGS_GRPC_ADDR": "127.0.0.1:10185",
              "SETTINGS_HTTP_ADDR": "127.0.0.1:10186",
              "SHARING_DEBUG_ADDR": "127.0.0.1:10151",
              "SHARING_GRPC_ADDR": "127.0.0.1:10150",
              "SSE_DEBUG_ADDR": "127.0.0.1:10135",
              "STORAGE_PUBLICLINK_DEBUG_ADDR": "127.0.0.1:10179",
              "STORAGE_PUBLICLINK_GRPC_ADDR": "127.0.0.1:10178",
              "STORAGE_SHARES_DEBUG_ADDR": "127.0.0.1:10156",
              "STORAGE_SHARES_GRPC_ADDR": "127.0.0.1:10154",
              "STORAGE_SYSTEM_DATA_SERVER_URL": "http://localhost:10216/data",
              "STORAGE_SYSTEM_DEBUG_ADDR": "127.0.0.1:10217",
              "STORAGE_SYSTEM_GRPC_ADDR": "127.0.0.1:10215",
              "STORAGE_SYSTEM_HTTP_ADDR": "127.0.0.1:10216",
              "STORAGE_USERS_DEBUG_ADDR": "127.0.0.1:10159",
              "STORAGE_USERS_GRPC_ADDR": "127.0.0.1:10157",
              "STORAGE_USERS_HTTP_ADDR": "127.0.0.1:10158",
              "STORAGE_USERS_DATA_SERVER_URL": "http://localhost:10158/data",
              "STORE_DEBUG_ADDR": "127.0.0.1:10464",
              "STORE_GRPC_ADDR": "127.0.0.1:10460",
              "THUMBNAILS_DEBUG_ADDR": "127.0.0.1:10194",
              "THUMBNAILS_GRPC_ADDR": "127.0.0.1:10191",
              "THUMBNAILS_HTTP_ADDR": "127.0.0.1:10190",
              "USERLOG_DEBUG_ADDR": "127.0.0.1:10210",
              "USERS_DEBUG_ADDR": "127.0.0.1:10145",
              "USERS_GRPC_ADDR": "127.0.0.1:10144",
              "WEB_DEBUG_ADDR": "localhost:10104",
              "WEB_HTTP_ADDR": "localhost:10100",
              "WEBDAV_DEBUG_ADDR": "127.0.0.1:10119",
              "WEBDAV_HTTP_ADDR": "127.0.0.1:10115",
              "OCIS_SERVICE_ACCOUNT_ID": "serviceaccount",
              "OCIS_SERVICE_ACCOUNT_SECRET": "serviceaccountsecret",
              "OCIS_DECOMPOSEDFS_PROPAGATOR": "async",
              "STORAGE_USERS_ASYNC_PROPAGATOR_PROPAGATION_DELAY": "10s",
              "TLS_INSECURE": "true",
              "OCIS_INSECURE": "true",
              "IDP_ACCESS_TOKEN_EXPIRATION": "9000",
              "GATEWAY_STORAGE_USERS_MOUNT_ID": "storage-users-1",
              "STORAGE_USERS_MOUNT_ID": "storage-users-1",
              "MICRO_REGISTRY": "memory",
              "SHARING_PUBLIC_DRIVER": "jsoncs3",
              "FRONTEND_OCS_RESOURCE_INFO_CACHE_TTL": "0",
              "OCIS_LOG_LEVEL": "info",
              "OCIS_LOG_PRETTY": "true",
              "OCIS_LOG_COLOR": "true",
              "PROXY_ENABLE_BASIC_AUTH": "true",
              "IDM_CREATE_DEMO_USERS": "true",
              "OCIS_ADMIN_USER_ID": "some-admin-user-id-0000-000000000000", // FIXME currently must have the length of a UUID, see reva/pkg/storage/utils/decomposedfs/spaces.go:228
              "IDM_ADMIN_PASSWORD": "admin",
              "OCIS_SYSTEM_USER_ID": "some-system-user-id-000-000000000000", // FIXME currently must have the length of a UUID, see reva/pkg/storage/utils/decomposedfs/spaces.go:228
              "OCIS_SYSTEM_USER_API_KEY": "some-system-user-machine-auth-api-key",
              "OCIS_JWT_SECRET": "some-ocis-jwt-secret",
              "OCIS_MACHINE_AUTH_API_KEY": "some-ocis-machine-auth-api-key",
              "OCIS_TRANSFER_SECRET": "some-ocis-transfer-secret",
              "IDM_SVC_PASSWORD": "some-ldap-idm-password",
              "GRAPH_LDAP_BIND_PASSWORD": "some-ldap-idm-password",
              "IDM_REVASVC_PASSWORD": "some-ldap-reva-password",
              "GROUPS_LDAP_BIND_PASSWORD": "some-ldap-reva-password",
              "USERS_LDAP_BIND_PASSWORD": "some-ldap-reva-password",
              "AUTH_BASIC_LDAP_BIND_PASSWORD": "some-ldap-reva-password",
              "IDM_IDPSVC_PASSWORD": "some-ldap-idp-password",
              "IDP_LDAP_BIND_PASSWORD": "some-ldap-idp-password",
              "GRAPH_APPLICATION_ID": "application-1"
          }
      },
```

</details>

- open ocis. go to 
image

### CODE REVIEW

File: .drone.star [hunk 1]
```diff
@@ -2039,7 +2059,11 @@ def ocisServer(storage, accounts_hash_difficulty = 4, volumes = [], depends_on =
         environment["APP_PROVIDER_WOPI_APP_URL"] = "http://fakeoffice:8080"
         environment["APP_PROVIDER_WOPI_INSECURE"] = "true"
         environment["APP_PROVIDER_WOPI_WOPI_SERVER_EXTERNAL_URL"] = "http://wopiserver:9300"
-        environment["APP_PROVIDER_WOPI_FOLDER_URL_BASE_URL"] = "https://ocis-server:9200"
+        environment["APP_PROVIDER_WOPI_FOLDER_URL_BASE_URL"] = OCIS_URL
+
+    if deploy_type == "federation-server":
```
Comment (saw-jan) [inline .drone.star:2064]: maybe we can just say federation type
```suggestion
    if deploy_type == "federation":
```

File: .drone.star [hunk 2]
```diff
@@ -2065,18 +2089,21 @@ def ocisServer(storage, accounts_hash_difficulty = 4, volumes = [], depends_on =
...
-        "name": "wait-for-ocis-server",
+        "name": "wait-for-%s" % (container_name),
         "image": OC_CI_ALPINE,
+        "environment": {
+            "OCIS_SERVER": container_name,
+        },
         "commands": [
             # wait for ocis-server to be ready (5 minutes)
-            "timeout 300 bash -c 'while [ $(curl -sk -uadmin:admin https://ocis-server:9200/graph/v1.0/users/admin -w %{http_code} -o /dev/null) != 200 ]; do sleep 1; done'",
+            'timeout 300 bash -c \'while [ "$(curl -sk -uadmin:admin https://${OCIS_SERVER}:9200/graph/v1.0/users/admin -w %{http_code} -o /dev/null)" != 200 ]; do sleep 1; done\'',
```
Comment (saw-jan) [inline .drone.star:2099]: you should be able to build the command string without using env
```suggestion
            "timeout 300 bash -c 'while [ $(curl -sk -uadmin:admin " +
            "%s/graph/v1.0/users/admin " % environment["OCIS_URL"] +
            "-w %{http_code} -o /dev/null) != 200 ]; do sleep 1; done'",
```

File: .drone.star [hunk 3]
```diff
@@ -2065,18 +2089,21 @@ def ocisServer(storage, accounts_hash_difficulty = 4, volumes = [], depends_on =
     ]
```
Comment (saw-jan) [inline .drone.star:2089]: wrapper command also needs to be updated.
```diff
- "%s/bin/ociswrapper serve --bin %s --url %s --admin-username admin --admin-password admin" % (dirs["ocisWrapper"], ocis_bin, OCIS_URL),
+ "%s/bin/ociswrapper serve --bin %s --url %s --admin-username admin --admin-password admin" % (dirs["ocisWrapper"], ocis_bin, environment["OCIS_URL"]),
```

File: tests/TestHelpers/OcmHelper.php [hunk 4]
```diff
@@ -0,0 +1,108 @@
+<?php declare(strict_types=1);
+/**
+ * ownCloud
+ *
+ * @author Kiran Parajuli <kiran@jankaritech.com>
+ * @copyright Copyright (c) 2022 Kiran Parajuli kiran@jankaritech.com
```
Comment (saw-jan) [inline tests/TestHelpers/OcmHelper.php:6]: .

File: tests/TestHelpers/OcmHelper.php [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+namespace TestHelpers;
+
+use GuzzleHttp\Exception\GuzzleException;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+
+/**
+ * A helper class for managing Graph API requests
+ */
+class OcmHelper {
```
Comment (saw-jan) [inline tests/TestHelpers/OcmHelper.php:18]:
```suggestion
/**
 * A helper class for managing federation server requests
 */
class OcmHelper {
```

File: tests/acceptance/features/apiOcm/ocm.feature [hunk 6]
```diff
@@ -0,0 +1,21 @@
+Feature: an user shares resources usin ScienceMesh application
+  As a user
+  I want to share resources between different ocis instances
+
+
+  Background:
```
Comment (saw-jan) [inline tests/acceptance/features/apiOcm/ocm.feature:6]: Background can have single newline
```suggestion

  Background:
```

File: tests/acceptance/features/bootstrap/OcmContext.php [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+ *
+ */
+
+use Behat\Behat\Context\Context;
+use Behat\Behat\Hook\Scope\BeforeScenarioScope;
+use GuzzleHttp\Exception\GuzzleException;
+use TestHelpers\OcisHelper;
+use TestHelpers\OcmHelper;
+
+require_once 'bootstrap.php';
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/OcmContext.php:28]: is it required?

File: tests/TestHelpers/OcmHelper.php [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+	public static function acceptInvitation(
+		string $baseUrl,
+		string $xRequestId,
+		string $user,
+		string $password,
+		string $token
+	): ResponseInterface {
+		$body = [
+		  "token" => $token,
+		  "providerDomain" => 'localhost'
```
Comment (saw-jan) [inline tests/TestHelpers/OcmHelper.php:96]: is `localhost` correct?

File: tests/acceptance/features/apiOcm/ocm.feature [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    Given these users have been created with default attributes and without skeleton files:
+      | username |
+      | Alice    |
+    And using server "REMOTE"
+    And user "Brian" has been created with default attributes and without skeleton files
+
+
+  Scenario: user generates invitation 
+    Given using server "LOCAL"
+    When "Alice" generates invitation
```
Comment (saw-jan) [inline tests/acceptance/features/apiOcm/ocm.feature:16]: can invitation be created without specifying any recipient?

File: tests/acceptance/features/bootstrap/OcmContext.php (hunk 7 above)
Comment (ScharfViktor) [reply tests/acceptance/features/bootstrap/OcmContext.php:28]: no, copy paste

File: tests/TestHelpers/OcmHelper.php (hunk 8 above)
Comment (ScharfViktor) [reply tests/TestHelpers/OcmHelper.php:96]: ah thanks, should be `ocis-server` for CI and `localhost` for local running.
I would take care of this in the next PR and add documentation on how to run ocm tests.

File: tests/acceptance/features/apiOcm/ocm.feature (hunk 9 above)
Comment (ScharfViktor) [reply tests/acceptance/features/apiOcm/ocm.feature:16]: Yes, here we are just creating a token. We can specify the user's email, but this is not a required field.

File: .drone.star [hunk 10]
```diff
@@ -2061,22 +2085,27 @@ def ocisServer(storage, accounts_hash_difficulty = 4, volumes = [], depends_on =
...
+        "%s/bin/ociswrapper serve --bin %s --url %s --admin-username admin --admin-password admin" % (dirs["ocisWrapper"], ocis_bin, environment["OCIS_URL"]),
     ]
 
     wait_for_ocis = {
-        "name": "wait-for-ocis-server",
+        "name": "wait-for-%s" % (container_name),
         "image": OC_CI_ALPINE,
+        "environment": {
+            "OCIS_SERVER": container_name,
+        },
```
Comment (saw-jan) [inline .drone.star:2096]: this can be removed now
```suggestion
```

**Review by saw-jan — APPROVED** [review]
others look good. :+1:

**Pull request merged**